src/Components/Components/src/RenderTree/EventArgsTypeCache.cs

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Components.RenderTree
+{
+    internal static class EventArgsTypeCache
+    {
+        private static ConcurrentDictionary<MethodInfo, Type> Cache = new ConcurrentDictionary<MethodInfo, Type>();
+
+        public static Type GetEventArgsType(MethodInfo methodInfo)
+        {
+            return Cache.GetOrAdd(methodInfo, methodInfo =>
+            {
+                var parameterInfos = methodInfo.GetParameters();
+                if (parameterInfos.Length == 0)
+                {
+                    return typeof(EventArgs);
+                }
+                else if (parameterInfos.Length > 1)
+                {
+                    throw new InvalidOperationException($"The method {methodInfo} cannot be used as an event handler because it declares more than one parameter.");
+                }
+                else
+                {
+                    var declaredType = parameterInfos[0].ParameterType;
+                    if (typeof(EventArgs).IsAssignableFrom(declaredType))
+                    {
+                        return declaredType;
+                    }
+                    else
+                    {
+                        throw new InvalidOperationException($"The event handler parameter type {declaredType.FullName} for event must inherit from {typeof(EventArgs).FullName}.");
```

Comment:
We don't really have to have this rule, but I'd err on the side of strictness. We can always loosen such rules later, but can't tighten them.

Inheriting from `EventArgs` is something we've made assumptions about in other places, plus it's good to make the developer really indicate for sure that they want to deserialize arbitrary untrusted input into this type.

Reply:
I think it's legit to force inheriting from EventArgs. As you mention, we need a gesture to mark these types as suitable for custom events. That said, I'm not 100% sure extending EventArgs is enough for this. Mainly because many types outside Blazor extend from it and we would want to avoid those "spurious" usages. 

Maybe we can mark it with an attribute instead to be more explicit. Specially since this has security implications in SSB

Reply:
The other gesture a developer does to opt into this is actually using the type as the parameter to an event handler definition (with `[EventHandler(...)]`). Having two indicators (subclassing `EventArgs` and explicitly putting it on one of these definitions) is a clear enough statement of intent IMO.

The risk is similar to a `[FromBody] SomeType myParam` on an MVC action method. In that case we consider it a sufficiently clear statement of intent with fewer steps and restrictions.